src/lua/protobuf.lua

```diff
@@ -0,0 +1,624 @@
+        max_key = math.max(max_key, k)
+    end
+    if key_count == 0 then
+        return
+    end
+    if min_key ~= 1 then
+        error(('Input array for %q repeated field got min index %d. ' ..
+            'Must be 1'):format(field_def.name, min_key))
+    end
+    if max_key ~= key_count then
+        error(('Input array for %q repeated field has inconsistent keys. ' ..
+            'Got table with %d fields and max index of %d'):format(
+            field_def.name, key_count, max_key))
+    end
+end
+
+local function validate_type(field_def, value, exp_type)
+    if type(exp_type) == 'table' then
+        local found = false
+        for _, exp_t in pairs(exp_type) do
+            if type(value) == exp_t then
+                found = true
+                break
+            end
+        end
+        if not found then
+            error(('Field %q of %q type gets %q type value.'):format(
+                field_def.name, field_def.type, type(value)))
+        end
+        return
+    end
+    assert(type(exp_type) == 'string')
+    if type(value) ~= exp_type then
+        error(('Field %q of %q type gets %q type value.'):format(
+            field_def.name, field_def.type, type(value)))
+    end
+    return
+end
+
+local function validate_value_error(field_def, value)
+    error(('Input data for %q field is %q and do not fit in %q')
+        :format(field_def.name, value, field_def.type))
+end
+
+local function validate_value(field_def, value, range)
+    local min = range ~= nil and range[1] or nil
+    local max = range ~= nil and range[2] or nil
+    if min ~= nil and value < min or max ~= nil and value > max then
+        validate_value_error(field_def, value)
+    end
+end
+
+validate_scalar = function(field_def, value)
+    local scalar_def = scalars[field_def.type]
+    assert(scalar_def.accept_type ~= nil)
+    validate_type(field_def, value, scalar_def.accept_type)
+    if scalar_def.accept_type == 'string' then
+        validate_length(value)
+    end
+    if type(value) == 'number' and is_nan(value) then
+        error(('Input data for %q field is NaN'):format(field_def.name))
+    end
+    if type(value) == 'number' and is_inf(value) then
+        error(('Input data for %q field is inf'):format(field_def.name))
+    end
+    if scalar_def.integral_only then
+        check_integer(field_def, value)
+    end
+    if scalar_def.limits ~= nil then
+        local value_type = type(value)
+        if value_type == 'cdata' then
+            value_type = ffi.istype(int64_t, value) and 'int64'
+                or 'uint64'
+        end
+        validate_value(field_def, value, scalar_def.limits[value_type])
+    end
+end
+
+-- }}} Validation
+
+
+-- {{{ Scalar type definitions
+
+scalars.float = {
+    accept_type = 'number',
+    encode_as_packed = true,
+    integral_only = false,
+    limits = {
+        number = {-3.4028234E+38, 3.4028234E+38},
+    },
+    encode = wireformat.encode_float,
+}
```

Comment:
I'm worrying that these limits are not precise. For example, the maximum valid 32-bit IEEE 754 floating point value is `0x1.fffffep+127`, but the given validation forbids it.

I've found that LuaJIT supports the `0x<...>p+<...>` literal format [from Lua 5.2](https://www.lua.org/manual/5.2/manual.html#3.1) and we can represent such values precisely. I propose to use literals of this kind in `scalars.float.limits.number`.

There is another option. We can do something like the following:

```lua
local function isfinite(v)
    local d = v - v
    return v == v and d == d
end

local function is_in_float_range(v)
    local float_p = ffi.new('float[1]')
    float_p[0] = v
    return isfinite(tonumber(float_p[0]))
end
```

As I see, `double` to `float` cast in C gives infinity for numbers beyond the float range. However, a value 'slightly' above maximum float is rounded to the maximum float. For example, double value `(2 - 2^(-23)) * 2^127 + 2^100` converted to float and then converted to double back is equal to the max float: `(2 - 2^(-23)) * 2^127`. It seems, it is related to the round-to-nearest semantic.

I don't know, whether it is good to follow this semantic or bad for our use case. I also don't know, whether this behavior is guaranteed.

Given this, I would keep the range check, but use the hexadecimal literals for accuracy.

CC @Buristan
CC @locker

Reply:
Replaced number values in range borders with hexadecimal values. 

At the same time I was thinking do we need to check float values for underflow when the number value is too small for float to represent?

Reply:
Too small by modulo (close to zero)? I would discard values outside of the 32-bit float range (roughly [-3.4E+38; 3.4E+38]) and accept ones within the range.

It means that we proceed with the values within the range with a precision loss, including values that are not equal to zero as 64-bit double, but equal to zero as 32-bit float.

My assumption is that one who use `float` in the protobuf schema is ready to get values with `float`'s precision.

Reply:
Ok, got it left only the overflow check